helix-view/src/env.rs

```diff
@@ -0,0 +1,13 @@
+use std::fmt::Display;
+
+use crate::Document;
+
+pub fn inject_environment<T>(strs: T, doc: &Document) -> Vec<String>
+where
+    T: Iterator,
+    <T as Iterator>::Item: Display,
+{
+    let path = doc.path().and_then(|p| p.to_str()).unwrap_or_default();
+
+    strs.map(|s| s.to_string().replace("$file", path)).collect()
+}
```

Comment:
I'm not sure this is possible because of lifetimes (trait items need static lifetimes when they are return values) but it would be nice if this returned something implementing `IntoIterator`, to prevent collecting since [`Command::args`](https://doc.rust-lang.org/std/process/struct.Command.html#method.args) takes an IntoIterator. It might work to return a `impl Iterator<String>` and remove the `.collect()`?

Reply:
```patch
diff --git a/helix-term/src/commands.rs b/helix-term/src/commands.rs
index b943d6d4..4a2a48f2 100644
--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -4586,12 +4586,10 @@ fn shell_impl(
     use std::process::{Command, Stdio};
     ensure!(!shell.is_empty(), "No shell set");
 
-    let cmd = inject_environment(args.iter(), doc).join(" ");
-
     let mut process = Command::new(&shell[0]);
     process
         .args(&shell[1..])
-        .arg(cmd)
+        .args(inject_environment(args.iter(), doc))
         .stdout(Stdio::piped())
         .stderr(Stdio::piped());
 
diff --git a/helix-view/src/env.rs b/helix-view/src/env.rs
index 4db32308..678e26a1 100644
--- a/helix-view/src/env.rs
+++ b/helix-view/src/env.rs
@@ -2,7 +2,10 @@
 
 use crate::Document;
 
-pub fn inject_environment<T>(strs: T, doc: &Document) -> Vec<String>
+pub fn inject_environment<'a, T: 'a>(
+    strs: T,
+    doc: &'a Document,
+) -> impl Iterator<Item = String> + 'a
 where
     T: Iterator,
     <T as Iterator>::Item: Display,
@@ -20,6 +23,6 @@ pub fn inject_environment<T>(strs: T, doc: &Document) -> Vec<String>
             "ENV:NO_PATH".to_string()
         });
 
-    strs.map(|s| s.to_string().replace("$file", path).replace("$pwd", &pwd))
-        .collect()
+    strs.into_iter()
+        .map(move |s| s.to_string().replace("$file", path).replace("$pwd", &pwd))
 }
```
The code compiles, but now the feature no longer works. Ideas? (I'm in way over my head on this one, still learning)

Reply:
The switch from `.arg` to `.args` is what's doing it I believe: the command should be interpreted like `sh -c "echo foo"` but this switches it to `sh -c echo foo` since it's no longer one arg.

Looking at the callsites though I'm not sure returning an iterator saves any work. A Vec might be better here actually since Vec's have `Vec::join` and iterators don't and that's how this is usually consumed.

Reply:
Should we close this conversation as resolved then?